### PR TITLE
Add atop connector

### PR DESCRIPTION
**Pull request by cberner**

### CODE REVIEW

File: presto-atop/src/main/java/com/facebook/presto/atop/AtopTable.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+                long millis = Long.valueOf(fields.get(7));
+                type.writeLong(builder, millis);
+            }),
+            new AtopColumn("read_requests", StandardTypes.BIGINT, bigintParser(8)),
+            new AtopColumn("sectors_read", StandardTypes.BIGINT, bigintParser(9)),
+            new AtopColumn("write_requests", StandardTypes.BIGINT, bigintParser(10)),
+            new AtopColumn("sectors_written", StandardTypes.BIGINT, bigintParser(11))
+    )),
+
+    REBOOTS("reboots", "DSK", ImmutableList.of(AtopColumn.HOST_IP, new AtopColumn("power_on_time", StandardTypes.TIMESTAMP, (fields, type, builder) -> {
```
Comment (martint) [inline presto-atop/src/main/java/com/facebook/presto/atop/AtopTable.java]: power_on_time should probably be timestamp with time zone

File: presto-atop/src/main/java/com/facebook/presto/atop/AtopConnectorId.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.atop;
+
+import java.util.Objects;
+
+import static java.util.Objects.requireNonNull;
+
+public final class AtopConnectorId
```
Comment (dain) [inline presto-atop/src/main/java/com/facebook/presto/atop/AtopConnectorId.java]: You shouldn't need connector ids anymore

File: presto-atop/src/main/java/com/facebook/presto/atop/AtopMetadata.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+            return null;
+        }
+    }
+
+    @Override
+    public List<ConnectorTableLayoutResult> getTableLayouts(ConnectorSession session,
+                                                            ConnectorTableHandle table,
+                                                            Constraint<ColumnHandle> constraint,
+                                                            Optional<Set<ColumnHandle>> desiredColumns)
+    {
```
Comment (dain) [inline presto-atop/src/main/java/com/facebook/presto/atop/AtopMetadata.java]: formatting

File: presto-atop/src/main/java/com/facebook/presto/atop/AtopMetadata.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        if (schemaNameOrNull == null) {
+            return Stream.of(AtopTable.values())
+                    .map(table -> new SchemaTableName(environment, table.getName()))
+                    .collect(Collectors.toList());
+        }
+        if (!listSchemaNames(session).contains(schemaNameOrNull)) {
+            return ImmutableList.of();
+        }
+        return Stream.of(AtopTable.values())
+                .map(table -> new SchemaTableName(schemaNameOrNull, table.getName()))
```
Comment (dain) [inline presto-atop/src/main/java/com/facebook/presto/atop/AtopMetadata.java]: does this need to be lower case?

File: presto-atop/src/main/java/com/facebook/presto/atop/AtopPageSourceProvider.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+            AtopColumnHandle atopColumnHandle = checkType(column, AtopColumnHandle.class, "column");
+            AtopColumn atopColumn = atopSplit.getTable().getColumn(atopColumnHandle.getName());
+            atopColumns.add(atopColumn);
+            types.add(typeManager.getType(atopColumn.getType()));
+        }
+
+        // Timezone is not preserved during JSON serialization
+        DateTime date = atopSplit.getDate().withZone(timeZone);
+        checkState(date.equals(date.withTimeAtStartOfDay()), "Expected date to be at beginning of day");
+        return new AtopPageSource(readerPermits, atopFactory, Slices.utf8Slice(atopSplit.getHost().getHostText()), atopSplit.getTable(), date, atopColumns.build(), types.build());
```
Comment (dain) [inline presto-atop/src/main/java/com/facebook/presto/atop/AtopPageSourceProvider.java]: static import

File: presto-atop/src/main/java/com/facebook/presto/atop/AtopProcessFactory.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        this.executablePath = config.getExecutablePath();
+        this.timeZone = config.getDateTimeZone();
+    }
+
+    @Override
+    public Atop create(AtopTable table, DateTime date)
+    {
+        checkArgument(date.getZone().equals(timeZone), "Split date (%s) is not in the local timezone (%s)", date.getZone(), timeZone);
+
+        ProcessBuilder processBuilder = new ProcessBuilder(executablePath);
```
Comment (dain) [inline presto-atop/src/main/java/com/facebook/presto/atop/AtopProcessFactory.java]: Consider using Airlift Command library to manage the process (https://github.com/airlift/command/blob/master/src/main/java/io/airlift/command/Command.java).  There are many edge cases in managing process from Java.

File: presto-atop/src/main/java/com/facebook/presto/atop/AtopTable.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+import static com.facebook.presto.atop.AtopTable.AtopColumn.START_TIME;
+import static com.facebook.presto.atop.AtopTable.AtopColumnParser.bigintParser;
+import static com.facebook.presto.atop.AtopTable.AtopColumnParser.varcharParser;
+import static com.facebook.presto.spi.type.TypeSignature.parseTypeSignature;
+import static java.util.Objects.requireNonNull;
+
+public enum AtopTable
+{
+    DISKS("disks", "DSK", baseColumnsAnd(
+            new AtopColumn("device_name", StandardTypes.VARCHAR, varcharParser(6)),
```
Comment (dain) [inline presto-atop/src/main/java/com/facebook/presto/atop/AtopTable.java]: Static import

File: presto-atop/src/main/java/com/facebook/presto/atop/AtopTableHandle.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        }
+        AtopTableHandle other = (AtopTableHandle) obj;
+        return Objects.equals(this.schema, other.schema) &&
+                Objects.equals(this.table, other.table);
+    }
+
+    @Override
+    public String toString()
+    {
+        return Joiner.on(":").join(schema, table);
```
Comment (dain) [inline presto-atop/src/main/java/com/facebook/presto/atop/AtopTableHandle.java]: I'd just use `schema + ':' + table`

**Comment by dain** [comment]
Other than some minor comments, the process management is my only real concern.

File: presto-atop/src/main/java/com/facebook/presto/atop/AtopConnectorId.java (hunk 2 above)
Comment (cberner) [inline presto-atop/src/main/java/com/facebook/presto/atop/AtopConnectorId.java]: I do for logging

File: presto-atop/src/main/java/com/facebook/presto/atop/AtopMetadata.java (hunk 4 above)
Comment (cberner) [inline presto-atop/src/main/java/com/facebook/presto/atop/AtopMetadata.java]: It doesn't look like the Hive connector lower cases them, and mine are lower case anyway (it's just the label for passing to atop that's upper case)

**Comment by cberner** [comment]
Added two commits to address the comments. Will squash them before merging

File: presto-atop/src/main/java/com/facebook/presto/atop/AtopProcessFactory.java [hunk 9]
```diff
@@ -63,30 +77,42 @@ public Atop create(AtopTable table, DateTime date)
...
         private String line;
 
-        private AtopProcess(Process process)
+        private AtopProcess(Process process, Duration readTimeout, ExecutorService executor)
         {
             this.process = requireNonNull(process, "process is null");
-            reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
+            underlyingReader = new BufferedReader(new InputStreamReader(process.getInputStream()));
+            TimeLimiter limiter = new SimpleTimeLimiter(executor);
+            this.reader = limiter.newProxy(underlyingReader::readLine, LineReader.class, readTimeout.toMillis(), MILLISECONDS);
```
Comment (dain) [inline presto-atop/src/main/java/com/facebook/presto/atop/AtopProcessFactory.java]: Can you add a test that hangs and test that the code cleans up correctly?

**Comment by dain** [comment]
Github is being a bit buggy, so I'll repeat my comment: Can you add a test that hangs and test that the code cleans up correctly?

**Comment by dain** [comment]
Looks good

**Pull request merged**